### PR TITLE
ArduinoCLI preprocessing

### LINKED ISSUE #14: Support for .ino/.pde library-less projects
Wrap [~~`arduino-preprocessor`~~](https://github.com/arduino/arduino-preprocessor) (discarded since too unstable) [`arduino-cli`](https://github.com/arduino/arduino-cli) to preprocess `.ino` projects and forward them to the normal C++ toolchain

### PR DESCRIPTION
**Pull request by AeroStun**
Closes #14

### CODE REVIEW

File: modules/ardrivo/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        src/entrypoint.cxx)
+target_compile_definitions (Ardrivo INTERFACE)
+target_link_libraries (Ardrivo INTERFACE stdpolyfills::stdpolyfills)
+
+#add_executable (TestsArdrivo)
+#set_target_properties (TestsArdrivo PROPERTIES LINKER_LANGUAGE CXX)
+#target_link_libraries (TestsArdrivo Ardrivo Catch2)
+
+#catch_discover_tests (TestsArdrivo)
+#add_dependencies(TestTag TestsArdrivo)
```
Comment (RuthgerD) [inline modules/ardrivo/CMakeLists.txt:32]: Unneeded commented out cmake. (Implement tests or remove)

File: modules/trycompile/src/Toolchain.cxx [hunk 2]
```diff
@@ -48,8 +54,10 @@ CompilationResults compile_sketch(SketchSource src, stdfs::path prefix) {
     constexpr std::string_view silent_output = "null.txt"; // just in case
 #endif
 
-    const auto configure_cmd = fmt::format(R"(cmake -DSOURCE_FILE="{0}" -S "{1}" -B "{2}" > {3} && cmake --build "{4}" > {3} 2<&1)",
-                                           src.location.string(), share_dir.string(), build_dir.string(), silent_output, build_dir.string());
+
+    const auto configure_cmd = fmt::format(R"(cmake -DSOURCE_FILE="{0}" -DSOURCE_TYPE="{5}" -DARDUINOCLI_PATH="{6}" -S "{1}" -B "{2}" > {3} && cmake --build "{4}" > {3} 2<&1)",
```
Comment (RuthgerD) [inline modules/trycompile/src/Toolchain.cxx:58]: We might need 2<&1 on the first cmake call as well (I'm pretty sure it was there in the first place but got lost due to debugging stuff)

File: share/CMakeLists.txt [hunk 3]
```diff
@@ -46,4 +72,4 @@ set (CMAKE_POSITION_INDEPENDENT_CODE ON)
 #target_include_directories (SmartcarShield PUBLIC ${SMARTCAR_SOURCE_PATH})
 #target_link_libraries (SmartcarShield PUBLIC ArdRuntime)
```
Comment (RuthgerD) [inline share/CMakeLists.txt:73]: Some more unused commented cmake

File: modules/ardrivo/src/entrypoint.cxx [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include "entrypoint.hxx"
+
+bool init(void*) { return true; }
```
Comment (RuthgerD) [inline modules/ardrivo/src/entrypoint.cxx:21]: I need those newlines at the end of the file :)

**Comment by AeroStun** [comment]
Trailing newlines are not part of the style guidelines of this project and their absence is not specific to this PR.
If you would like to add them, please edit the Wiki to include a section dedicated to project-specific guidelines and open a PR against `master` to homogenize the trailing newline behavior

File: share/Modules/Preproc.cmake [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+#
+
+function (identify_source SOURCE_FILE SOURCE_TYPE)
+    set (SOURCE_FILE_ORIGINAL "${SOURCE_FILE}" PARENT_SCOPE)
+    if (SOURCE_TYPE STREQUAL "sketch")
+        if (SOURCE_FILE MATCHES "\\.ino$" OR SOURCE_FILE MATCHES "\\.pde$" )
+            set (SOURCE_IS_CPP OFF PARENT_SCOPE)
+            set (SOURCE_IS_CPP OFF)
+            set (SOURCE_IS_INO ON PARENT_SCOPE)
+            set (SOURCE_IS_INO ON)
```
Comment (SyncInProgress) [inline share/Modules/Preproc.cmake:25]: It looks like the code is repeated in lines 32-35 and 39-42. If there isn't a way to change it, not a problem. Everything else looks pretty good.

Comment (AeroStun) [reply share/Modules/Preproc.cmake:25]: The only ways to deduplicate these wouldn't make the code any more sane to read (quite the opposite I'm afraid)

File: include/ardpolyfills/Arduino.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/*
```
Comment (idreessamadi) [inline include/ardpolyfills/Arduino.h:1]: don't we need to add the file name here?

Comment (AeroStun) [reply include/ardpolyfills/Arduino.h:1]: Solved in 095e0795

File: modules/ardrivo/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#
```
Comment (davidos-b) [inline modules/ardrivo/CMakeLists.txt:1]: #  CMakeLists.txt
#  Copyright 2020 ItJustWorksTM
?????

Comment (AeroStun) [reply modules/ardrivo/CMakeLists.txt:1]: Took me more than a minute to figure out what you meant; I'd suggest learning about Markdown formatting
https://guides.github.com/features/mastering-markdown/

Comment (AeroStun) [reply modules/ardrivo/CMakeLists.txt:1]: Fixed in a1f2040 (from 9f35925)

Comment (davidos-b) [reply modules/ardrivo/CMakeLists.txt:1]: Hah yeah, in case anybody is without glasses or something

**Pull request merged**